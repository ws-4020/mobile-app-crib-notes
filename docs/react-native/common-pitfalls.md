---
title: ハマりがちな落とし穴
sidebar_label: Overview
---

React Nativeで開発する際に起きた困り事や不具合についてまとめています。
