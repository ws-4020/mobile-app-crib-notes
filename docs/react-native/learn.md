---
title: About
hide_table_of_contents: true
---

このコンテンツはReact Nativeでのモバイルアプリケーション開発を学習するためのコンテンツです。
