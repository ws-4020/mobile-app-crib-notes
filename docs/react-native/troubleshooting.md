---
title: トラブルシュート
---

React Nativeで開発する中で起きた困り事や不具合についてまとめています。
