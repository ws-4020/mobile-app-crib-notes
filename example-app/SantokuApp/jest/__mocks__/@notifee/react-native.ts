/**
 * Copyright 2023 TIS Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import notifee from '@notifee/react-native';
export * from '@notifee/react-native/dist/types/Library';
export * from '@notifee/react-native/dist/types/Notification';
export * from '@notifee/react-native/dist/types/Trigger';
export * from '@notifee/react-native/dist/types/NotificationIOS';
export * from '@notifee/react-native/dist/types/NotificationAndroid';

const mock: typeof notifee = {
  cancelAllNotifications: jest.fn(),
  cancelDisplayedNotifications: jest.fn(),
  cancelTriggerNotifications: jest.fn(),
  cancelNotification: jest.fn(),
  cancelDisplayedNotification: jest.fn(),
  cancelTriggerNotification: jest.fn(),
  createChannel: jest.fn(),
  createChannels: jest.fn(),
  createChannelGroup: jest.fn(),
  createChannelGroups: jest.fn(),
  deleteChannel: jest.fn(),
  deleteChannelGroup: jest.fn(),
  displayNotification: jest.fn(),
  openAlarmPermissionSettings: jest.fn(),
  createTriggerNotification: jest.fn(),
  getTriggerNotificationIds: jest.fn(),
  getDisplayedNotifications: jest.fn(),
  getTriggerNotifications: jest.fn(),
  getChannel: jest.fn(),
  isChannelCreated: jest.fn(),
  isChannelBlocked: jest.fn(),
  getChannels: jest.fn(),
  getChannelGroup: jest.fn(),
  getChannelGroups: jest.fn(),
  getInitialNotification: jest.fn(),
  onBackgroundEvent: jest.fn(),
  onForegroundEvent: jest.fn(),
  openNotificationSettings: jest.fn(),
  registerForegroundService: jest.fn(),
  stopForegroundService: jest.fn(),
  requestPermission: jest.fn(),
  setNotificationCategories: jest.fn(),
  getNotificationCategories: jest.fn(),
  getNotificationSettings: jest.fn(),
  getBadgeCount: jest.fn(),
  setBadgeCount: jest.fn(),
  incrementBadgeCount: jest.fn(),
  decrementBadgeCount: jest.fn(),
  openBatteryOptimizationSettings: jest.fn(),
  isBatteryOptimizationEnabled: jest.fn(),
  getPowerManagerInfo: jest.fn(),
  openPowerManagerSettings: jest.fn(),
  hideNotificationDrawer: jest.fn(),
  SDK_VERSION: '5.4.1',
};

// 複数のファイルでnotifeeをimportしていた場合に、redefineされないようにモックが存在するか確認
if (!__mocks.notifee) {
  Object.defineProperty(__mocks, 'notifee', {value: mock});
}

export default mock;
