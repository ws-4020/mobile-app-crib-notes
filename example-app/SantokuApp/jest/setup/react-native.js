/**
 * Copyright 2023 TIS Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// https://reactnavigation.org/docs/testing#mocking-native-modules
// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// !!! NOTICE !!!
// The React Navigation documentation says that `react-native-reanimated` and `react-native-gesture-handler`
// also need to be configured with mocks.
// But since React Navigation v6 does not already depend on those packages, mocking those packages is not necessary
// for React Navigation v6.
//
// If you want to use `react-native-reanimated` and `react-native-gesture-handler`, add mock according to the guide
// provided by each package.
// - `react-native-reanimated`: https://docs.swmansion.com/react-native-reanimated/docs/guide/testing/
// - `react-native-gesture-handler`: https://docs.swmansion.com/react-native-gesture-handler/docs/guides/testing
