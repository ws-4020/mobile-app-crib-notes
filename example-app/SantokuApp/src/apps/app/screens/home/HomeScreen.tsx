import {useFocusEffect} from '@react-navigation/core';
import {HomePage} from 'features/home/components/HomePage';
import {useShowTermsAgreementOverlayUseCase} from 'features/terms/hooks/useShowTermsAgreementOverlayUseCase';
import React from 'react';

export const HomeScreen: React.FC = () => {
  // 利用規約に未同意の場合は、利用規約を表示します。
  const showTermsAgreementOverlay = useShowTermsAgreementOverlayUseCase();
  useFocusEffect(showTermsAgreementOverlay);

  return <HomePage />;
};
