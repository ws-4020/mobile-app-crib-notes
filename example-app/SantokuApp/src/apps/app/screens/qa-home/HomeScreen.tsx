import {CompositeScreenProps} from '@react-navigation/native';
import {NativeStackNavigationOptions, NativeStackScreenProps} from '@react-navigation/native-stack';
import {StatusBar, setStatusBarStyle} from 'expo-status-bar';
import {HomePage} from 'features/qa-home/pages/HomePage';
import React, {useCallback, useEffect} from 'react';

import {AuthenticatedStackParamList, HomeStackParamList} from '../../navigators/types';

export const HomeScreen: React.FC<
  CompositeScreenProps<
    NativeStackScreenProps<HomeStackParamList, 'Home'>,
    NativeStackScreenProps<AuthenticatedStackParamList, 'QuestionAndEventStackNav'>
  >
> = ({navigation}) => {
  const navigateToQuestionDetail = useCallback(
    (questionId: string) => navigation.navigate('QuestionDetail', {questionId}),
    [navigation],
  );

  const navigateToQuestionAndEventPost = useCallback(
    () => navigation.navigate('QuestionAndEventStackNav', {screen: 'QuestionAndEventPost'}),
    [navigation],
  );

  const setNavigationOptions = useCallback(
    (options: NativeStackNavigationOptions) => navigation.setOptions(options),
    [navigation],
  );

  useEffect(() => {
    const listener = navigation.addListener('focus', () => {
      setStatusBarStyle('light');
    });
    return listener;
  }, [navigation]);

  return (
    <>
      <StatusBar style="light" />
      <HomePage
        navigateToQuestionAndEventPost={navigateToQuestionAndEventPost}
        navigateToQuestionDetail={navigateToQuestionDetail}
        setHeader={setNavigationOptions}
      />
    </>
  );
};
