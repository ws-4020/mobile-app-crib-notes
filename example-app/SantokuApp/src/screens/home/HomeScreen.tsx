import {useNavigation} from '@react-navigation/native';
import React, {useCallback, useContext} from 'react';
import {ThemeContext} from 'react-native-elements';

import {HomeTemplate} from './HomeTemplate';

const useNavigate = (screen: string) => {
  const navigation = useNavigation();
  return useCallback(() => navigation.navigate(screen), [navigation, screen]);
};

const ScreenName = 'Home';
const Screen: React.FC = () => {
  const {theme} = useContext(ThemeContext);
  const onGoToInstructionButtonPress = useNavigate('Instructions');
  const onViewApplicationInformationButtonPress = useNavigate('AppInfo');
  const onGoToDemoButtonPress = useNavigate('Demo');

  return (
    <HomeTemplate
      testID="HomeScreen"
      theme={theme}
      onGoToInstructionButtonPress={onGoToInstructionButtonPress}
      onViewApplicationInformationButtonPress={onViewApplicationInformationButtonPress}
      onGoToDemoButtonPress={onGoToDemoButtonPress}
    />
  );
};

// Navigatorに登録する情報
export const HomeScreen = {
  component: Screen,
  name: ScreenName,
  options: {
    headerShown: false,
  },
};
