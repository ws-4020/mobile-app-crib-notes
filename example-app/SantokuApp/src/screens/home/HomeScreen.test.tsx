import '@testing-library/jest-native/extend-expect';
import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';

import {HomeScreen} from './HomeScreen';

beforeEach(() => {
  jest.useFakeTimers();
});

describe('Home', () => {
  it('マウントされたときに正常にレンダリングされること', () => {
    const app = render(<HomeScreen.component />);
    expect(app.queryByTestId('HomeScreen')).toHaveTextContent('Hello, World!');
    expect(app).toMatchSnapshot();
  });

  it('Demoをタップすると、デモ画面に遷移すること', () => {
    const app = render(<HomeScreen.component />);
    fireEvent.press(app.getByTestId('goToDemoButton'));
    expect(__mocks.navigation.navigate).toHaveBeenCalledTimes(1);
    expect(__mocks.navigation.navigate).toHaveBeenCalledWith('DemoStackNav', {screen: 'Demo'});
  });

  it('Go to instructionsをタップすると、Reactの説明画面に遷移すること', () => {
    const app = render(<HomeScreen.component />);
    fireEvent.press(app.getByTestId('goToInstructionsButton'));
    expect(__mocks.navigation.navigate).toHaveBeenCalledTimes(1);
    expect(__mocks.navigation.navigate).toHaveBeenCalledWith('DemoStackNav', {screen: 'Instructions'});
  });

  it('Application Informationをタップすると、アプリ情報画面に遷移すること', () => {
    const app = render(<HomeScreen.component />);
    fireEvent.press(app.getByTestId('viewApplicationInformationButton'));
    expect(__mocks.navigation.navigate).toHaveBeenCalledTimes(1);
    expect(__mocks.navigation.navigate).toHaveBeenCalledWith('DemoStackNav', {screen: 'AppInfo'});
  });

  it('コンポーネントがマウントされてから200ms経ってからスプラッシュスクリーンが閉じられること', () => {
    render(<HomeScreen.component />);

    jest.advanceTimersByTime(100);
    expect(__mocks.expoSplashScreen.hideAsync).not.toHaveBeenCalled();

    // 200ms経過してから、スプラッシュスクリーンを閉じる
    jest.advanceTimersByTime(200);
    // FIXME: 呼び出される回数が安定しないので、呼ばれたことだけ検証する
    expect(__mocks.expoSplashScreen.hideAsync).toHaveBeenCalled();
  });

  it('スプラッシュスクリーンを閉じるのに失敗しても何もしないこと', () => {
    __mocks.expoSplashScreen.hideAsync.mockImplementationOnce(() =>
      Promise.reject(new Error('Splash screen is already closed')),
    );

    const app = render(<HomeScreen.component />);
    expect(__mocks.expoSplashScreen.hideAsync).not.toHaveBeenCalled();

    jest.advanceTimersByTime(100);
    expect(__mocks.expoSplashScreen.hideAsync).not.toHaveBeenCalled();

    jest.advanceTimersByTime(200);
    // FIXME: 呼び出される回数が安定しないので、呼ばれたことだけ検証する
    expect(__mocks.expoSplashScreen.hideAsync).toHaveBeenCalled();

    expect(app).toMatchSnapshot();
  });
});
