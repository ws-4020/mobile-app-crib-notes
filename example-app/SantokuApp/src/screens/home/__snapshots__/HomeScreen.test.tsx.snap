// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home スプラッシュスクリーンを閉じるのに失敗しても何もしないこと 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#2089dc",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="HomeScreen"
>
  <Text
    style={
      Object {
        "color": "#fefefe",
        "fontSize": 51.42857142857143,
      }
    }
    testID="HelloWorld"
  >
    Hello, World!
  </Text>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="goToInstructionsButton"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Go to instructions
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="viewApplicationInformationButton"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          View application information
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="goToDemoButton"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Demo
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Home マウントされたときに正常にレンダリングされること 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="HomeScreen"
>
  <Text
    style={
      Object {
        "color": "#242424",
        "fontSize": 43.714285714285715,
      }
    }
  >
    開発中
  </Text>
</View>
`;
